massa-ledger/src/cursor.rs

```diff
@@ -0,0 +1,125 @@
+use massa_hash::{Hash, HashDeserializer};
+use massa_models::{address::AddressDeserializer, Address};
+use massa_serialization::{Deserializer, SerializeError, Serializer};
+use nom::IResult;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LedgerCursorStep {
+    Start,
+    Balance,
+    Bytecode,
+    Datastore(Option<Hash>),
+    Finish,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct LedgerCursorStepSerializer;
+
+impl LedgerCursorStepSerializer {
+    fn _new() -> Self {
+        Self
+    }
+}
+
+impl Serializer<LedgerCursorStep> for LedgerCursorStepSerializer {
+    fn serialize(&self, value: &LedgerCursorStep) -> Result<Vec<u8>, SerializeError> {
+        match value {
+            LedgerCursorStep::Start => Ok(vec![0]),
+            LedgerCursorStep::Balance => Ok(vec![1]),
+            LedgerCursorStep::Bytecode => Ok(vec![2]),
+            LedgerCursorStep::Datastore(key) => {
+                let mut bytes = vec![3];
+                if let Some(key) = key {
+                    bytes.extend(key.to_bytes());
+                }
+                Ok(bytes)
+            }
+            LedgerCursorStep::Finish => Ok(vec![4]),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct LedgerCursorStepDeserializer;
+
+impl LedgerCursorStepDeserializer {
+    fn _new() -> Self {
+        Self
+    }
+}
+
+impl Deserializer<LedgerCursorStep> for LedgerCursorStepDeserializer {
+    fn deserialize<'a>(&self, bytes: &'a [u8]) -> IResult<&'a [u8], LedgerCursorStep> {
+        match bytes[0] {
+            0 => Ok((&bytes[1..], LedgerCursorStep::Start)),
+            1 => Ok((&bytes[1..], LedgerCursorStep::Balance)),
+            2 => Ok((&bytes[1..], LedgerCursorStep::Bytecode)),
+            3 => {
+                let hash_deserializer = HashDeserializer::new();
+                if bytes[1..].is_empty() {
+                    Ok((&bytes[1..], LedgerCursorStep::Datastore(None)))
+                } else {
+                    let (rest, key) = hash_deserializer.deserialize(&bytes[1..])?;
+                    Ok((rest, LedgerCursorStep::Datastore(Some(key))))
+                }
+            }
+            4 => Ok((&bytes[1..], LedgerCursorStep::Finish)),
+            _ => Err(nom::Err::Error(nom::error::Error::new(
+                bytes,
+                nom::error::ErrorKind::Digit,
+            ))),
+        }
+    }
+}
```

Comment:
I believe you can associate values to enum fields in rust, something like: 
```rust
enum MyEnum {
    A = 1,
    B = 2,
    C = 3,
}
```

And convert a byte directly to it's enum correspondant with https://docs.rs/num-derive/0.3.3/num_derive/

Reply:
I tested your crate and this one : https://github.com/illicitonion/num_enum which is more maintained and both, sadly, don't works with enums with fields on it like my field `Datastore` :/

Reply:
That's kinda sad